### PR TITLE
Locks can be acquired recursively with some versions of mysql

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
From MariaDB 10.0.2 it is possible to recursively set the same lock. This completely breaks the locking mechanism in `CRM_Core_Lock` because it assumes that `GET_LOCK` can only be called once for a named lock.

See: https://mariadb.com/kb/en/get_lock/ "From MariaDB 10.0.2, it is also possible to recursively set the same lock".

This is certainly the case on MariaDB 10.3. But I ran some tests on 10.5 and it doesn't seem to allow you to recursively set the same lock anymore.  So I think this goes a long way to explaining why locking doesn't seem to be working properly on some sites.

Before
----------------------------------------
CiviCRM locking mechanism doesn't work on versions of MariaDB (between 10.0.2-10.4.x?) as multiple locks of the same name can be acquired.

After
----------------------------------------
CiviCRM locking mechanism works on mariaDB to limit each lock to a single named lock.

Technical Details
----------------------------------------
Described above.

Comments
----------------------------------------
@seamuslee001 @johntwyman

### CODE REVIEW

**Comment by seamuslee001** [comment]
Test fails may relate here, but lets run again to be sure, Jenkins re test this please

**Comment by seamuslee001** [comment]
ok test failures replicate 

```
api_v3_JobProcessMailingTest.testConcurrency with data set #1
api_v3_JobProcessMailingTest.testConcurrency with data set #2
api_v3_JobProcessMailingTest.testConcurrency with data set #3
api_v3_JobProcessMailingTest.testConcurrency with data set #5
Civi.FlexMailer.ConcurrentDeliveryTest.testConcurrency with data set #1
Civi.FlexMailer.ConcurrentDeliveryTest.testConcurrency with data set #2
Civi.FlexMailer.ConcurrentDeliveryTest.testConcurrency with data set #3
Civi.FlexMailer.ConcurrentDeliveryTest.testConcurrency with data set #5
```

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
@seamuslee001 I'm not really sure what's going on with this now `Error: Maximum function nesting level of &#039;256&#039; reached, aborting! in substr() (line 653 of /home/jenkins/bknix-dfl/build/core-22013-24s1d/web/sites/all/modules/civicrm/vendor/pear/db/DB.php).` any ideas?

**Comment by Stoob** [comment]
Very helpful thanks @mattwire

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Looks like this is still failing tests & isn't reviewable - lets' close for now & you can re-open when you want to get it reviewable

**Comment by jmcclelland** [comment]
Hi - I'm not sure why the tests were failing, but I think this PR solves a problem. The LOCK mechanism in MariaDB seems to be designed to prevent one client from stepping on a second client, but it does not stop one client from stepping on itself. In the [documentation referenced by @mattwire](https://mariadb.com/kb/en/get_lock/):

> It is also possible to recursively set the same lock. If a lock with the same name is set n times, it needs to be released n times as well. 

Which implies that a lock with the same name _can_ be set over and over again. And...I just ran a test (MariaDB 10.5):

```
MariaDB [outreach]> SELECT GET_LOCK('lock1',10);
+----------------------+
| GET_LOCK('lock1',10) |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.001 sec)

MariaDB [outreach]> SELECT GET_LOCK('lock1',10);
+----------------------+
| GET_LOCK('lock1',10) |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.001 sec)

MariaDB [outreach]>
```
In other words, for the same client, you _can_ get as many locks as you want over and over again. The MySQL [documentation](https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html) seems to suggest that MySQL behaves the same way.

In contracts, I _think_ our `CRM_Core_Lock` class intends to prevent this behavior - and this [issue](https://lab.civicrm.org/dev/core/-/issues/3988) seems to be a case where it might be failing to fulfill this goal.

Lastly... I think @mattwire is a fix, but seems to be vulnerable to a race condition - since a lock could be acquired between checking and setting. I'm not sure what a better solution would be.

**Comment by eileenmcnaughton** [comment]
@jmcclelland if the issue is within the SAME process then it looks like the problem is we are only php tracking like this

```
    if (stristr($this->_name, 'data.mailing.job.')) {
          self::$jobLog = $this->_id;
        }
```

We probably should be tracking an array of held locks in `civi::statics` - although it rather feels like we might be leaning on the lock to do more than it was ever intended to do if we are expecting it ti stop recursion in the same process

**Comment by mattwire** [comment]
> We probably should be tracking an array of held locks in civi::statics

In my experience this won't work. The kind of race conditions that we were seeing with eg. smartgroup caching were happening too fast for civi::statics - ie. processes were probably executing the same line at the same time so a variable in civi::statics would simultaneously be not set and be set at the same time - a bit like Schroedingers cat..

The use of SELECT IS_USED_LOCK and SELECT GET_LOCK in this PR avoids that problem because (I believe) that mysql takes care of the concurrency issue.

**Comment by mattwire** [comment]
I should also note I use this patch on all my client sites for over a year now with no issues.

**Comment by eileenmcnaughton** [comment]
" ie. processes were probably executing the same line at the same time so a variable in civi::statics would simultaneously be not set and be set at the same time" - I thought the point of this PR was that the SAME process was aquiring the lock recurstively - but you are actually talking about 2 separate processes?

**Comment by jmcclelland** [comment]
I think either `SELECT IS_USED_LOCK()` or `civi::statics` would be an improvement over what we currently have. And technically, both of them are vulnerable to a race condition. 

For us to have a problem, the two requests don't have to happen at the exact same time, just within the same second. If both requests come one millisecond apart - it seems that both of them are going to get the same "NO" answer from either `SELECT IS_USED_LOCK()` or `civi::statics`.

I'm out of my depth here... but I'll venture to say: `SELECT IS_USED_LOCK()`  doesn't require a line of code to remove the index from the `civi::statics` since the lock goes away when it is released. And it's re-assuring to know that it's been tested in production. But it _seems_ that `civi::statics` would be faster than `SELECT IS_USED_LOCK()` since it doesn't require a trip to the database and back so might be preferable?

**Comment by jmcclelland** [comment]
Also, the SQL lock should protect us from two processes trying to acquire the same lock - so I think we are only adding code to address the problem of the same process acquiring a lock that it previously acquired.

**Comment by eileenmcnaughton** [comment]
@jmcclelland ok - so if the process is the same process recurstively aquiring the same lock to do the same update - how does that occur? Because the backtrace on the issue didn't demonstrate that happening. The only 'story' I could tell that would explain it involved delayed commits due to transactions - which would be unchanged by this?

The locking mechanism is generally intended to prevent 2 separate processes clashing rather than recursion within one process

**Comment by jmcclelland** [comment]
Ah - thanks @eileenmcnaughton for the clarification. I was under the impression that recursion is happening and that's the problem we are trying to solve for [this problem](https://lab.civicrm.org/dev/core/-/issues/3988), but I don't have a story for how that's happening given the backtraces provided.  I also didn't full comprehend what you meant by delayed commits due to transactions, but now I think I get what you are saying:

1. Process A starts transaction
2. Get's lock
3. INSERTS to cache (but it's delayed)
4. Releases lock
5. Process B starts transaction
6. Get's lock (since it's been released)
7. INSERTS to cache (but it's delay)
8. Process A completes transaction, INSERT is made
9. Process B completes transaction, INSERT is made, constraint is violated

That seems like a more plausible story for [#398](https://lab.civicrm.org/dev/core/-/issues/3988).

This PR still may be relevant because recursion may be happening else where. 

I'll follow up on 398 with this clarification.

**Comment by eileenmcnaughton** [comment]
I just opened https://lab.civicrm.org/dev/core/-/issues/4193 so we can close this without losing track of it until we have agreement as to what it is solving & how. We did find a fix for the problem @jmcclelland was working on - which I suspect might have had wide-ranging symptoms

I am going to make a note about it in the dev digest
